src/messages/serialize.rs

```diff
@@ -0,0 +1,64 @@
+//! Serialization rules specified in [RFC-001#serializationdeserialization]
+//!
+//! We automatically serialize and deserialize using serde derivations where possible.
+//! Sometimes we need to implement ourselves, this file holds that code.
+//!
+//! [RFC-001#rules]: https://github.com/ZcashFoundation/redjubjub/blob/main/rfcs/0001-messages.md#serializationdeserialization
+
+use serde::ser::{Serialize, Serializer};
+
+use serde::de::{self, Deserialize, Deserializer, Visitor};
+
+use super::constants::{AGGREGATOR_PARTICIPANT_ID, DEALER_PARTICIPANT_ID};
+use super::*;
+
+use std::fmt;
+
+impl Serialize for ParticipantId {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        match *self {
+            ParticipantId::Signer(id) => serializer.serialize_u8(id),
+            ParticipantId::Dealer => serializer.serialize_u8(DEALER_PARTICIPANT_ID),
+            ParticipantId::Aggregator => serializer.serialize_u8(AGGREGATOR_PARTICIPANT_ID),
+        }
+    }
+}
+
+struct ParticipantIdVisitor;
+
+impl<'de> Visitor<'de> for ParticipantIdVisitor {
+    type Value = ParticipantId;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter
+            .write_str(format!("an integer between {} and {}", std::u8::MIN, std::u8::MAX).as_str())
+    }
+
+    // We need to use u64 instead of u8 here because the JSON deserialized will call
+    // `visit_u64` for any unsigned int:
+    // https://serde.rs/impl-deserialize.html#driving-a-visitor
+    fn visit_u64<E>(self, value: u64) -> Result<Self::Value, E>
```

Comment:
Does this force users to use JSON?
Do you need to implement both `visit_u64` and `visit_u8` here, so users can use JSON and non-JSON serializers?

Reply:
I did some serialization using `bincode` and it is still using the `visit_u64`. I will add the `visit_u8` if we find a deserializer that actually uses it.